src/reanimated2/animation/spring.ts

```diff
@@ -93,39 +155,125 @@ export function withSpring(
+      animation: InnerSpringAnimation,
+      config: Partial<SpringConfig> &
+        Required<
+          Pick<SpringConfig, 'restSpeedThreshold' | 'restDisplacementThreshold'>
+        >,
+      current: number
+    ): {
+      isOvershooting: boolean;
+      isVelocity: boolean;
+      isDisplacement: boolean;
+    } {
+      const { toValue, velocity } = animation;
 
-      const isOvershooting = () => {
-        if (config.overshootClamping && config.stiffness !== 0) {
-          return current < toValue
+      const isOvershooting =
+        config.overshootClamping && config.stiffness !== 0
+          ? current < toValue
             ? animation.current > toValue
-            : animation.current < toValue;
-        } else {
-          return false;
-        }
-      };
+            : animation.current < toValue
+          : false;
 
       const isVelocity = Math.abs(velocity) < config.restSpeedThreshold;
       const isDisplacement =
         config.stiffness === 0 ||
         Math.abs(toValue - current) < config.restDisplacementThreshold;
 
-      if (zeta < 1) {
-        animation.current = underDampedPosition;
-        animation.velocity = underDampedVelocity;
-      } else {
-        animation.current = criticallyDampedPosition;
-        animation.velocity = criticallyDampedVelocity;
+      return { isOvershooting, isVelocity, isDisplacement };
+    }
+
+    function calcuateNewMassToMatchDuration(x0: number) {
+      /** Use this formula: https://phys.libretexts.org/Bookshelves/University_Physics/Book%3A_University_Physics_(OpenStax)/Book%3A_University_Physics_I_-_Mechanics_Sound_Oscillations_and_Waves_(OpenStax)/15%3A_Oscillations/15.06%3A_Damped_Oscillations
+           * to find the asympotote and esitmate the damping that gives us the expected duration 
+
+                ⎛ ⎛ c⎞           ⎞           
+                ⎜-⎜──⎟ ⋅ duration⎟           
+                ⎝ ⎝2m⎠           ⎠           
+            A ⋅ e                   = treshold
+
+     
+          Amplitude calculated using "Conservation of energy"
+                           _________________
+                          ╱      2         2
+                         ╱ m ⋅ v0  + k ⋅ x0 
+          amplitude =   ╱  ─────────────────
+                      ╲╱           k        
+
+          And replace mass with damping ratio which is provided: m = (c^2)/(4 * k * zeta^2)   
+          */
+      const {
+        duration,
+        velocity: v0,
+        stiffness: k,
+        dampingRatio: zeta,
+        restSpeedThreshold: threshold,
+      } = config;
+
+      const durationForMass = (mass: number) => {
+        const amplitude = Math.sqrt((mass * v0 * v0 + k * x0 * x0) / k);
+        const c = zeta * 2 * Math.sqrt(k * mass);
+        return (
+          1000 * ((-2 * mass) / c) * Math.log((threshold * 0.01) / amplitude) -
+          duration
+        );
+      };
+
+      return bisectRoot({ min: 0, max: 50, func: durationForMass });
```

Comment:
Perhaps I'll add here additional comment before merging - usually this is not the best way to find a root and sb in the future may want to improve it.